### PR TITLE
Enable mimalloc on pytorch Windows

### PR DESCRIPTION
**Pull request by xuhancn**
This PR is implemention of [#102534](https://github.com/pytorch/pytorch/issues/102534), option 2.
Major changes:
1. Add mimalloc to the submodule.
2. Add build option "USE_MIMALLOC". 
3. It is only enabled on Windows build, And it would improve pytorch memory allocation performance.

Additional Test:
image
This PR also build & static link mimalloc on Linux well.

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -348,6 +348,11 @@ cmake_dependent_option(
 cmake_dependent_option(
     BUILD_FUNCTORCH "Build Functorch" ON "BUILD_PYTHON" OFF)
 
+option(USE_MIMALLOC "Use mimalloc" OFF)
+# Enable third party mallic library to improve memory allocation performance on Windows.
```
Comment (jgong5) [inline CMakeLists.txt:352]:
```suggestion
# Enable third party mimalloc library to improve memory allocation performance on Windows.
```

**Review by jgong5 — APPROVED** [review]
Please add the issue id that this PR addresses in the description.

File: CMakeLists.txt (hunk 1 above)
Comment (xuhancn) [reply CMakeLists.txt]: done

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**